lib/CL/devices/remote/communication.c

```diff
@@ -1480,57 +1467,54 @@ find_or_create_server (const char *address_with_port, unsigned port,
   d->fast_port = port;
   d->slow_port = port + 1;
 
-  if (pocl_network_connect (d, &d->fast_socket_fd, d->fast_port,
-                            NETWORK_BUF_SIZE_FAST, 1))
-    {
-      POCL_MSG_ERR ("Could not connect to server\n");
-      POCL_MEM_FREE (d);
-      return NULL;
-    }
-
-  if (pocl_network_connect (d, &d->slow_socket_fd, d->slow_port,
-                            NETWORK_BUF_SIZE_SLOW, 0))
-    {
-      POCL_MSG_ERR ("Could not connect to server\n");
-      POCL_MEM_FREE (d);
-      return NULL;
-    }
-
-  ClientHandshake_t hs, hsr;
-  memset (&hs, 0, sizeof (ClientHandshake_t));
 #ifdef ENABLE_RDMA
   // TODO: re-enable once client RDMA has been reworked to match server
   // communication
   if (CL_TRUE || rdma_init_id (&d->rdma_data) == 0)
     {
-      hs.rdma_supported = 0;
+      // hs.m.get_session.use_rdma = 0;
     }
   else
     {
       POCL_MSG_ERR ("Could not create RDMAcm event channel and id, continuing"
                     "without RDMA\n");
     }
 #endif
-  ssize_t readb, writeb;
-  writeb = write (d->fast_socket_fd, &hs, sizeof (ClientHandshake_t));
-  assert ((size_t)(writeb) == sizeof (ClientHandshake_t));
-  readb = read (d->fast_socket_fd, &hsr, sizeof (ClientHandshake_t));
-  assert ((size_t)(readb) == sizeof (ClientHandshake_t));
-  memcpy (d->session, hsr.session_id, SESSION_ID_LENGTH);
+
+  ReplyMsg_t hsr;
+  if (pocl_network_connect (d, &d->fast_socket_fd, d->fast_port,
+                            NETWORK_BUF_SIZE_FAST, 1, &hsr))
+    {
+      POCL_MSG_ERR ("Could not connect to server\n");
+      POCL_MEM_FREE (d);
+      return NULL;
+    }
+
+  memcpy (d->authkey, hsr.m.get_session.authkey, AUTHKEY_LENGTH);
+  d->session = hsr.m.get_session.session;
 
   {
-    char tmp[2 * SESSION_ID_LENGTH + 1] = "................................";
-    for (uint8_t *b = d->session, i = 0; i < SESSION_ID_LENGTH; ++i)
+    char tmp[2 * AUTHKEY_LENGTH + 1] = "................................";
+    for (uint8_t *b = d->authkey, i = 0; i < AUTHKEY_LENGTH; ++i)
       snprintf (tmp + (2 * i), 3, "%02x", *b++);
-    POCL_MSG_PRINT_REMOTE ("Received session id %s\n", tmp);
+    POCL_MSG_PRINT_REMOTE ("Received session id %" PRIu64 ", key %s\n",
+                           hsr.m.get_session.session, tmp);
   }
 
-  d->peer_port = hsr.peer_port;
+  d->peer_port = hsr.m.get_session.peer_port;
+
+  if (pocl_network_connect (d, &d->slow_socket_fd, d->slow_port,
+                            NETWORK_BUF_SIZE_SLOW, 0, NULL))
+    {
+      POCL_MSG_ERR ("Could not connect to server\n");
+      POCL_MEM_FREE (d);
+      return NULL;
+    }
 
   DL_APPEND (servers, d);
 
 #ifdef ENABLE_RDMA
-  d->use_rdma = hsr.rdma_supported;
+  d->use_rdma = 0; // hsr.m.create_session.rdma_supported;
```

Comment:
Is this change on purpose? Disables RDMA?

Reply:
This only disables RDMA between the client and server, which has been broken for ages and has been forced off earlier in the code for an almost equally long time.